text/3857-cfg-version.md

```diff
@@ -0,0 +1,635 @@
+- The needs of users on the latest toolchain that expect integration with new features (e.g. `Error` in `core`) or faster compile times (dropping `is-terminal` dep in favor of `std::io::IsTerminal`)
+- The expectations they have set with their policy
+
+For a simple case, like `Error` in `core`,
+maintainers want to conditionally add the `impl core::error::Error` if its supported.
+
+In cases like `std::io::IsTerminal`,
+maintainers need to trim dependencies in Cargo for newer Rust versions to maintain reasonable build times for users on newer toolchains.
+
+A challenge with this is that in order to solve this,
+we need to add a new feature that requires waiting for an MSRV bump before it can be used.
+Being able to check for the presence of this feature would allow immediate adoption.
+
+### Testing proc-macros
+
+*(non-motivating)*
+
+Requires: programmatic, vendor version
+- Can be approximated by using the language version
+
+Error reporting can be a major usability issue for proc-macros.
+Packages like [`trybuild`](https://crates.io/crates/trybuild) exist to demonstrate and track
+the quality of errors reported by proc-macros by compiling sample code and snapshotting the compiler output.
+However, compiler output is dependent on the vendor and changes from release to release, so maintainers need to restrict the tests to specific Rust versions.
+
+For example, in `clap`'s [`derive_ui`](https://github.com/clap-rs/clap/blob/master/tests/derive_ui.rs) test:
+```rust
+#[cfg(feature = "derive")]
+#[rustversion::attr(not(stable(1.89)), ignore)] // STABLE
+#[test]
+fn ui() {
+    let t = trybuild::TestCases::new();
+    t.compile_fail("tests/derive_ui/*.rs");
+}
+```
+
+### Working around compiler bugs
+
+*(non-motivating)*
+
+Requires: programmatic, vendor version
+- Can be approximated by using the language version
+
+At times, a vendor's compiler has  bugs that need to be worked around,
+e.g. see [error-chain#101](https://github.com/rust-lang-deprecated/error-chain/issues/101).
+
+### Build information
+
+*(non-motivating)*
+
+Requires: display, vendor version
+- Can be approximated by using the language version
+
+Some applications choose to include build information in their verbose-version or `--bugreport`.
+This can include the compiler vendor and version used to build the application.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When using a new language or standard library feature,
+circumstances may warrant doing so while maintaining an existing [MSRV](https://doc.rust-lang.org/cargo/reference/rust-version.html),
+rather than raising to what the language or standard library feature needs.
+This can be accomplished by conditionally compiling the code for that feature.
+
+For instance, say you have an MSRV of 1.10 and `#[cfg(since)]` feature was available in 1.20, you would have been able to do:
+```rust
+#[cfg_attr(since(rust, "1.27"), must_use)]
```

Comment:
Text says 1.20 but the example uses 1.27.

Reply:
`#[must_use]` however is from 1.27.

Tried to clarify this with ee1c8cfc